libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/FileAccessTree.java

```diff
@@ -16,26 +16,49 @@
 import java.util.Arrays;
 import java.util.List;
 import java.util.Objects;
+import java.util.function.Consumer;
 
 import static org.elasticsearch.core.PathUtils.getDefaultFileSystem;
 
 public final class FileAccessTree {
-    public static final FileAccessTree EMPTY = new FileAccessTree(FilesEntitlement.EMPTY);
+
+    public static final FileAccessTree EMPTY = new FileAccessTree(FilesEntitlement.EMPTY, null);
     private static final String FILE_SEPARATOR = getDefaultFileSystem().getSeparator();
 
     private final String[] readPaths;
     private final String[] writePaths;
 
-    private FileAccessTree(FilesEntitlement filesEntitlement) {
+    private static void resolvePath(FilesEntitlement.FileData fileData, PathLookup pathLookup, Consumer<Path> resolvedPathReceiver) {
```

Comment:
Could we avoid the indirection of the "receiver" by having the resolve method on FileData? So instead of `fileData.path()` you have `fileData.resolvePaths(pathLookup)`. The members can be internal, FileData could even be an interface with 2 (soon to be 3 with setting paths) sealed impls. All FileAccessTree cares about is a path and a mode. Note that I explicitly used plural above, because with eg data dirs we could resolve to multiple paths.

Reply:
Sure, I see what you are saying here. Will do that. 
Anything else missing @rjernst?

Reply:
Turned out quite neat I think